OpenBSD support

jg1uaa opened the pull request:
simply two changes needed
- use default compiler
- add #include <signal.h>
On modern Linux, signal(2) works as BSD signal (not SysV). So simply add missed header.

jg1uaa commented:
How do we treat for (Open)Solaris/OpenIndiana?
I think this use real SystemV signal()... rewrite code with sigaction()??

dmsc commented:
Hi!

Thanks for the pull-request.

> How do we treat for (Open)Solaris/OpenIndiana? I think this use real SystemV signal()... rewrite code with sigaction()??

I don't have those for testing, but "signal" is in POSIX, C90 and C99, so it should be supported by all recent unices.

Have Fun!

jg1uaa commented:
Hello, thanks for merging.

> I don't have those for testing, but "signal" is in POSIX, C90 and C99, so it should be supported by all recent unices.

It is the problem that the behavior is different even if the name "signal" is same.
Please see the manpage of signal(2) for Linux (https://man7.org/linux/man-pages/man2/signal.2.html), Notes:Portability section.

At least *modern* Linux and *BSD's signal() works same, so this PR used very simple solution. But if someone needs to run emu2 on ancient Linux or (Open)Solaris/OpenIndiana, it will be the time to fix.

dmsc commented:
Hi!

Do you think this will work? https://github.com/dmsc/emu2/commit/bb0a2e92ef036a369053c6a57892117043754aff

Any way to test?

Have Fun!

jg1uaa commented:
I created OpenIndiana virtual machine by VMware Player. I tried to build, but there is a error about cfmakeraw() in src/keyb.c (not found). 
https://github.com/mitogen-hq/mitogen/issues/71 and https://perkin.org.uk/posts/solaris-portability-cfmakeraw.html will be a hint to implement, but I think this will make your code ugly.

If there is no strong reason to support OpenIndiana, I think the code should be unchanged including sigaction().

Nevertheless, using sigaction()...

https://github.com/jg1uaa/emu2/commit/81e465d7a8790f43fecc76838c69fe0095c507cc

- sighandler_t is GNU (Linux) extension, replaced with more portable definition
- there is no need to use sigaddset() without SA_NODEFER,  removed.
- replaced from 0 to NULL in sigaction()

I simply tested emu2 by running LHa.exe (archiver) on OpenBSD, Linux(void). If there is your recommended DOS program for testing emu2, please tell me.

dmsc commented:
Hi!

> I created OpenIndiana virtual machine by VMware Player. I tried to build, but there is a error about cfmakeraw() in src/keyb.c (not found). [mitogen-hq/mitogen#71](https://github.com/mitogen-hq/mitogen/issues/71) and https://perkin.org.uk/posts/solaris-portability-cfmakeraw.html will be a hint to implement, but I think this will make your code ugly.
>

The problem is that I don't know if all the flags that needs to be set to make the terminal "RAW" are available on sys-v, but I checked at least the flags available in POSIX, this should do it (in the same branch): https://github.com/dmsc/emu2/commit/ad8cc3ce15edb5130de3cdcf2f40387223195544

I also included your patch in my branch.

>If there is no strong reason to support OpenIndiana, I think the code should be unchanged including sigaction().

I tried downloading an OpenSolaris VM image, but could not start it under qemu, so I can't test it now.

> 
> I simply tested emu2 by running LHa.exe (archiver) on OpenBSD, Linux(void). If there is your recommended DOS program for testing emu2, please tell me.

You can test with QBASIC.EXE or other full-screen app.

Have Fun!

jg1uaa commented:
I checked compiling emu2 on following OSes.

- (Linux) void linux, Debian11/amd64
- (OmniOS) OmniOS r151040
- (OpenBSD) OpenBSD-7.0/amd64

OpenIndiana is too heavy for my PC so I use [OmniOS](https://omnios.org/).

emu2 + QBASIC.exe runs on these environments, but I found function key and Alt-modifier is not working.

| | TERMCAP | function key | Alt + ? |
|-- |-- |--| --|
|Linux console |linux| not working | working |
|OmniOS console |(not defined) |not working| not working|
|OpenBSD console |vt220|working |not working|
|mlterm| xterm | working | not working |
|xfce4-terminal|xterm-256color|working | not working|

I think this problem is caused by window manager or terminal app, *not emu2*.

- I use mlterm on OpenBSD with JWM. This window manager captures alt modifier, such as Alt+F4 (close window).
- xfce4-terminal itself captures Alt key for various menu.

What window manager and terminal emulator is good for test?

dmsc commented:
Hi!

Thank you for all your testing.

> I checked compiling emu2 on following OSes.
> 
> * (Linux) void linux, Debian11/amd64
> * (OmniOS) OmniOS r151040
> * (OpenBSD) OpenBSD-7.0/amd64
> 
> OpenIndiana is too heavy for my PC so I use [OmniOS](https://omnios.org/).
> 
> emu2 + QBASIC.exe runs on these environments, but I found function key and Alt-modifier is not working.
> 

Yes, many terminals don't pass all key combinations to the application. Here I use gnome-terminal, it has an option to disable the processing of the ALT-key for the menus.

Linux console uses different sequences for the F* keys, I could add support for those as there is no overlap with the standard xterm ones.

Have Fun!
